smack-experimental/src/main/java/org/jivesoftware/smackx/message_markup/provider/MarkupElementProvider.java

```diff
@@ -0,0 +1,140 @@
+/**
+ *
+ * Copyright © 2018 Paul Schaub
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jivesoftware.smackx.message_markup.provider;
+
+import static org.xmlpull.v1.XmlPullParser.END_TAG;
+import static org.xmlpull.v1.XmlPullParser.START_TAG;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.jivesoftware.smack.SmackException;
+import org.jivesoftware.smack.provider.ExtensionElementProvider;
+import org.jivesoftware.smack.util.ParserUtils;
+import org.jivesoftware.smackx.message_markup.MessageMarkup;
+import org.jivesoftware.smackx.message_markup.element.BlockQuoteElement;
+import org.jivesoftware.smackx.message_markup.element.CodeBlockElement;
+import org.jivesoftware.smackx.message_markup.element.ListElement;
+import org.jivesoftware.smackx.message_markup.element.MarkupElement;
+import org.jivesoftware.smackx.message_markup.element.SpanElement;
+
+import org.xmlpull.v1.XmlPullParser;
+
+public class MarkupElementProvider extends ExtensionElementProvider<MarkupElement> {
+
+    @Override
+    public MarkupElement parse(XmlPullParser parser, int initialDepth) throws Exception {
+
+        MessageMarkup markup = MessageMarkup.getBuilder();
+
+        int spanStart = -1, spanEnd = -1;
+        Set<SpanElement.SpanStyle> spanStyles = new HashSet<>();
+
+        int listStart = -1, listEnd = -1;
+        List<ListElement.ListEntryElement> lis = new ArrayList<>();
+
+        while (true) {
+            int tag = parser.next();
+            String name = parser.getName();
+            int start, end;
+            switch (tag) {
+                case START_TAG:
+                    switch (name) {
+                        case BlockQuoteElement.ELEMENT:
+                            start = ParserUtils.getIntegerAttributeOrThrow(parser, BlockQuoteElement.ATTR_START,
```

Comment:
This pattern repeats. I wonder if it would be a good idea to create a static helper method `StartEnd paraseStartEndAttributes(XmlPullParser)`, where `StartEnd` is simply a class with the two start and end integers. Ideally we also have then a setter in the element class which takes this type.

Reply:
I'm not sure whether it is worth to create those methods just now. That snippet wouldn't be reused too often and I think It would overall make the code more complicated than it is. Right now you can clearly see, what happens.